### PR TITLE
Let's rewrite the test harness

### PR DESCRIPTION
**Pull request by cbezault**
This change is mostly cleanups, though it brings us more in line with libcxx's testing format in terms of style and functionality.

I also tried to decrease the size of the `TestConfig` and `STLTest` objects as much as possible since those are pickled and communicated to the worker processes via IPC. On my machine I saw a ~20% decrease in testing time with this change (~75mins to ~60mins), we'll see if that holds true on the test machines.

### CODE REVIEW

File: tests/std/tests/VSO_0000000_any_calling_conventions/custom_format.py
Comment (cbezault) [inline tests/std/tests/VSO_0000000_any_calling_conventions/custom_format.py:8]: I don't need `TestType`

File: tests/libcxx/lit.cfg
Comment (cbezault) [inline tests/libcxx/lit.cfg:1]: This whole file is unnecessary.

**Comment by cbezault** [comment]
@rpjohnst If you could make sure I didn't break your edg_drop machinery that would be much appreciated.

**Comment by cbezault** [comment]
@StephanTLavavej, I think the most important thing for you to check is to make sure I didn't mangle the new standard_library_header_units test.

File: tests/utils/stl/test/format.py
Comment (rpjohnst) [inline tests/utils/stl/test/format.py:133]: It doesn't look like `isenseRspPath` is ever defined here- the callers use a variable with the same name to add the `--write-isense-rsp` flag, but only in an `if` scope so it never reaches here.

**Review by rpjohnst — COMMENTED** [review]
I've noted what looks like *an* issue below, but I also can't seem to get the test harness to reach that point at all- I have `STL_EDG_DROP` set but I never see any `cl ... /dE--write-isense-rsp` or `edgcpfe_drop.exe` processes run. I've also tried adding print statements to `_handleIsenseRspFile` but don't see those in the output either (possibly swallowed by lit?).

File: tests/utils/stl/test/format.py
Comment (cbezault) [reply tests/utils/stl/test/format.py:133]: A lot of output gets (unhelpfully) swallowed by lit. Give running the tests with the `-a` option a shot.

Comment (rpjohnst) [reply tests/utils/stl/test/format.py:133]: Ahh, thanks- looks like `_handleIsenseRspFile` is getting skipped because the `if TestType.COMPILE in test.testType` branch is using `return` instead of `yield`.

Once I patch that to a `yield` and forward `isenseRspPath` to `_handleIsenseRspFile`, the next problem is that `_handleIsenseRspFile`'s `yield TestStep` line doesn't have `shared` in scope for `shared.execDir`. (Incidentally, the exception handler in `execute` is also failing to report this, because it still uses `lit_config` instead of `litConfig`.)

Once I also forward `shared` to `_handleIsenseRspFile` it seems to work, though I haven't looked at the custom formats that also call `_handleIsenseRspFile`.

Comment (cbezault) [reply tests/utils/stl/test/format.py]: Alright I switched it up a bit and I think it should work now.
The only question is around multi-TU tests though I don't think this was working for them in the first place.

**Comment by cbezault** [comment]
There are some tests missing.
I'm not sure why but this can't be merged until that is resolved. Other test runs see 26007 tests, I'm seeing 25897 here.

**Comment by ben-craig** [comment]
If we aren't supposed to run lit directly anymore, then the readme needs updating as well

**Review by ben-craig — COMMENTED** [review]
There's a lot of deleted files in here, which in general, is a think I like to see.  I'll have to figure out where my kernel changes need to go with this new layout though.

File: tests/utils/stl/test/format.py
Comment (rpjohnst) [inline tests/utils/stl/test/format.py:281]: It looks like this section is missing a call to `self.runStep`- the report shows the wrong value for `cmd` and the binary in `STL_EDG_DROP` is never run.

File: stl/inc/atomic
Comment (StephanTLavavej) [reply stl/inc/atomic:2468]: I believe this is from #1090 which probably should be merged first:

> Fix a bug in `std::atomic_init` found by new tests: P0083R2 deprecated `atomic_init` _and_ changed its effects to be equivalent to a relaxed store. We had implemented the deprecation, but missed the effect change.

File: tests/std/lit.site.cfg.in
Comment (cbezault) [reply tests/std/lit.site.cfg.in:18]: Things should go in the config file if the information is needed to actually run the test. Otherwise you should attempt to keep the information in the lit_config. Configs are embedded in test objects which are pickled and sent to worker processes to execute. Keeping the test object small do the pickling and transmission is fast is important in my experience.

File: tests/libcxx/expected_results.txt
Comment (cbezault) [reply tests/libcxx/expected_results.txt:70]: This got pulled in from another PR which is a prerequisite to this one, like Stephan said.

File: tests/std/tests/VSO_0000000_any_calling_conventions/custom_format.py
Comment (cbezault) [reply tests/std/tests/VSO_0000000_any_calling_conventions/custom_format.py:18]: These custom test formats derive from the main test format but change what the build/run steps look like.

File: tests/utils/stl/test/format.py
Comment (cbezault) [reply tests/utils/stl/test/format.py:281]: Fixed from the Yosemite backcountry!

**Comment by StephanTLavavej** [comment]
@cbezault I tried to merge `master` with the following merge conflict resolutions:

* `tests/libcxx/lit.cfg`
  + Added by both PRs, accepted this PR's version and discarded GH-1090's
* `tests/libcxx/lit.site.cfg.in`
  + Changed by both PRs, accepted this PR's version. Verified identical before and after merge. Unsure if this is correct (i.e. whether anything from GH-1090 should be preserved here).
* `tests/libcxx/CMakeLists.txt`
  + Both PRs made the same change, but this PR added an extra line; accepted this PR's version.

This failed with:

```
S:\GitHub\STL\out\build\x86>python tests\utils\stl-lit\stl-lit.py ..\..\..\llvm-project\libcxx\test -o testing_x86.log
stl-lit.py: S:\GitHub\STL\tests\utils\stl\test\config.py:19: note: Enabling Lit feature 'long_tests' as a result of parameter 'long_tests'
stl-lit.py: S:\GitHub\STL\tests\utils\stl\test\config.py:29: note: Enabling implicitly detected Lit features 'edg', 'locale.en_US.UTF-8', 'locale.fr_CA.ISO8859-1', 'locale.fr_FR.UTF-8', 'locale.ru_RU.UTF-8', 'locale.zh_CN.UTF-8', 'msvc', 'windows'
stl-lit.py: S:\GitHub\STL\llvm-project\llvm\utils\lit\lit\TestingConfig.py:99: fatal: unable to parse config file 'S:\\GitHub\\STL\\llvm-project\\libcxx\\test\\libcxx\\input.output\\filesystems\\lit.local.cfg', traceback: Traceback (most recent call last):
  File "S:/GitHub/STL/llvm-project/llvm\utils\lit\lit\TestingConfig.py", line 88, in load_from_path
    exec(compile(data, path, 'exec'), cfg_globals, None)
  File "S:\GitHub\STL\llvm-project\libcxx\test\libcxx\input.output\filesystems\lit.local.cfg", line 3, in <module>
    std_filesystem_tests = os.path.join(config.test_source_root, 'std', 'input.output', 'filesystems')
  File "C:\Users\stl\AppData\Local\Programs\Python\Python39\lib\ntpath.py", line 78, in join
    path = os.fspath(path)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

So I haven't pushed the merge. I observe that this is mentioning `test_source_root` which GH-1090 added to `tests/libcxx/lit.site.cfg.in`, but I don't know what's happening here.

**Comment by cbezault** [comment]
I think you made all the right choices but just forgot to re-run the configuration step.

**Comment by StephanTLavavej** [comment]
I didn't even know that there was a configuration step - thank you.

**Comment by StephanTLavavej** [comment]
This has *very* impressive performance - other PRs are taking ~39min for the x86/x64 checks, while this just took 23-24min.

**Comment by cbezault** [comment]
This isn't for this PR but I made a small patch for lit which batches up tests which the workers receive instead of needing to be handed a test one at a time and saw test times of ~50 mins on my personal computer. So there's still significant room for improvement.

File: tests/utils/stl/test/format.py
Comment (ben-craig) [inline tests/utils/stl/test/format.py:275]: each of these batches of runSteps are basically copy+paste+tweaked.  Can the duplication be extracted?

```
    def runSteps(self, stageName, steps, litConfig):
        report = stageName + ' steps:\n'
        for step in steps:
            cmd, out, err, rc = self.runStep(step, litConfig)

            if step.shouldFail and rc == 0:
                report += stageName + ' step succeeded unexpectedly.\n'
            elif rc != 0:
                report += stageName = ' step failed unexpectedly.\n'

            report += stl.util.makeReport(cmd, out, err, rc)
            if (step.shouldFail and rc == 0) or (not step.shouldFail and rc != 0):
                litConfig.note(report)
                return lit.Test.Result(failVar, report)

    def execute(self, test, litConfig):
        # ...
        self.runSteps('Build setup', buildSetupSteps, litConfig)
        self.runSteps('Build', buildSteps, litConfig)
        if litConfig.edg_drop:
            self.runSteps('Intellisense response', handleIsenseRspFile, litConfig)
        self.runSteps('Test setup', testSetupSteps, litConfig)
        self.runSteps('Test', testSteps, litConfig)
```

Comment (cbezault) [reply tests/utils/stl/test/format.py:275]: Sure, I was thinking of doing that eventually.

File: tests/utils/stl/test/format.py
Comment (ben-craig) [inline tests/utils/stl/test/format.py:324]: s/lit_config/litConfig

File: tests/libcxx/lit.site.cfg.in
Comment (StephanTLavavej) [inline tests/libcxx/lit.site.cfg.in:26]: Something is wrong with libcxx's expected_results (not specifically this line, just had to comment somewhere). I'm seeing:

```
S:\GitHub\STL\out\build\x86>python tests\utils\stl-lit\stl-lit.py ..\..\..\llvm-project\libcxx\test ..\..\..\tests\std ..\..\..\tests\tr1 -o testing_x86.log
[...]
Testing Time: 6625.84s
  Unsupported      :  1895
  Passed           : 22816
  Expectedly Failed:    32
  Failed           :  1203
```
It looks like most tests marked `FAIL` aren't being treated as expected failures, and tests marked `SKIPPED` aren't being skipped (in particular, `rethrow_if_nested.pass.cpp` is one of the tests that displays a crash dialog box on my machine).

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Looks good to me modulo the `expected_results` issue; I'll push fixes for the minor issues I noticed. Thanks!

File: tests/utils/stl/test/params.py
Comment (StephanTLavavej) [reply tests/utils/stl/test/params.py]: Reworded to stay below 120 columns.

File: tests/utils/stl/test/tests.py
Comment (StephanTLavavej) [inline tests/utils/stl/test/tests.py:122]: It looks like this line should be:
```suggestion
            self.result = Result(UNSUPPORTED, 'This test was marked as unsupported by a lit.cfg')
```
but I'm only 80% sure.

File: tests/utils/stl/test/format.py
Comment (StephanTLavavej) [inline tests/utils/stl/test/format.py:255]: Typo:
```suggestion
                        report += stageName + ' step succeeded unexpectedly.\n'
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: tests/utils/stl/test/tests.py
Comment (cbezault) [reply tests/utils/stl/test/tests.py:122]: It can go either way.

**Review by StephanTLavavej — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I encountered one test failure because my repo root, `S:\GitHub\STL`, combined with an exceptionally long libcxx path, combined with `/analyze`'s emission of a long-named XML file, exceeded the MAX_PATH limit. This made me realize that `/analyze` is emitting XML files that we never, ever want. I've pushed a commit to change all occurrences of plain `/analyze` to `/analyze:autolog-` which will still emit the warnings we want, but not the XML files. With this change, everything passes for x86:

```
Testing Time: 6095.88s
  Skipped          :   358
  Unsupported      :  1895
  Passed           : 22678
  Expectedly Failed:  1015
```

This doesn't change our `/analyze:only` options. I'll file an issue about that.

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [inline tests/utils/stl/test/tests.py:66]:
```suggestion
                filename.endswith(('.compile.pass.cpp', '.compile.fail.cpp')):
```

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [inline tests/utils/stl/test/tests.py:73]: Wait, so `.fail.cpp` means the same thing as `.compile.fail.cpp`, instead of having the `RUN` semantic like `.pass.cpp`? That seems like a nasty inconsistency that will silently create an absence of test coverage when misused, which is likely to happen given that `.fail.cpp` tests in libc++ are run.

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [inline tests/utils/stl/test/tests.py:163]: We should not be defining this. If a test depends on the value of a libc++-internal macro, it's a bug in the test that needs to be fixed. (Several occurrences.)

FWIW, grepping the current llvm-project submodule I find occurrences of `_LIBCPP_CONSTEXPR_AFTER_CXX17` only in product headers.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]
I'm totally breaking @StephanTLavavej's 17-attack PR merge combo. 🐼

File: tests/utils/stl/test/tests.py
Comment (cbezault) [reply tests/utils/stl/test/tests.py:117]: `.fail.cpp` tests are meant to be the same as `compile.fail.cpp` tests and will eventually all be replaced.

File: tests/utils/stl/test/tests.py
Comment (cbezault) [reply tests/utils/stl/test/tests.py:210]: Hmmm, when I was running the tests I was having failures without this. Will try again.

File: tests/std/lit.site.cfg.in
Comment (cbezault) [reply tests/std/lit.site.cfg.in:15]: Nope

File: tests/utils/stl/test/tests.py
Comment (cbezault) [reply tests/utils/stl/test/tests.py:117]: https://github.com/llvm/llvm-project/blob/5b30d9adc0536eee7fe0f164a550084916899acc/libcxx/utils/libcxx/test/format.py#L138

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [inline tests/utils/stl/test/tests.py:58]:
```suggestion
            return Result(UNSUPPORTED, "Test requires the following unavailable features: %s" % msg)
```

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [inline tests/utils/stl/test/tests.py:63]:
```suggestion
            return Result(UNSUPPORTED, "Test does not support the following features and/or targets: %s" % msg)
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: tests/utils/stl/test/tests.py
Comment (CaseyCarter) [reply tests/utils/stl/test/tests.py]: Aha! I thought this sounded familiar: https://github.com/llvm/llvm-project/commit/24ab5787b44435d0c7e805a8579753e48be9f5b9.

**Review by CaseyCarter — APPROVED** [review]
All good now!

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for this overhaul, improving test performance! This will both increase maintainer/contributor productivity :rocket: and decrease our costs :money_mouth_face:.

**Pull request merged**
